inference-engine/src/transformations/src/transformations/op_conversions/einsum_decomposition.cpp

```diff
@@ -0,0 +1,683 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "transformations/op_conversions/einsum_decomposition.hpp"
+
+#include <memory>
+#include <ngraph/opsets/opset7.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+#include <ngraph/rt_info.hpp>
+#include <ngraph/validation_util.hpp>
+#include <transformations/utils/utils.hpp>
+
+#include "itt.hpp"
+
+namespace {
+/// \brief      Check if the EinsumDecomposition transformation is applicable to a given Einsum.
+/// The transformation is applicable if input subscript does not have repeated labels and ellipsis.
+///
+/// \param      subscript          A subscript to check its format
+///
+/// \return     true - applicable, false - not applicable
+///
+bool is_subscript_applicable(const std::string& subscript) {
+    auto labels = ngraph::opset7::Einsum::extract_labels(subscript);
+    auto unique_labels = std::unordered_set<std::string>(labels.begin(), labels.end());
+    return std::find(labels.begin(), labels.end(), "...") == labels.end() && unique_labels.size() == labels.size();
+}
+
+/// \brief      Compute einsum_path for a given Einsum node meaning that the (pseudo-)optimal
+/// order of operands contraction in terms of performance and memory consumption
+///
+/// \param      einsum_node         An input Einsum node
+///
+/// \return     a vector of pairs with input indices assuming that the intermediate result is
+/// appended in the tail
+///
+std::vector<std::pair<size_t, size_t>> compute_einsum_path(std::shared_ptr<const ngraph::opset7::Einsum> einsum_node) {
```

Comment:
```suggestion
std::vector<std::pair<size_t, size_t>> compute_einsum_path(const std::shared_ptr<const ngraph::opset7::Einsum>& einsum_node) {
```

Reply:
We agreed to fix it in the scope of another PR for reference Einsum Implementation